docs/data-provider-integration.md

```diff
@@ -0,0 +1,83 @@
+# Guidance for data providers integrating into OpenSAFELY
+
+---8<-- 'includes/data-builder-danger-header.md'
+
+!!! warning
+
+    This page is not yet complete.
+
+    More complete guides on working with OpenSAFELY Contracts for
+    external data providers will eventually be documented more fully here.
+
+## Introduction
+
+For a data provider to offer new data for OpenSAFELY, there are two
+technical requirements:
+
+1. The data being offered must satisfy OpenSAFELY Contracts.
+2. The data provider's backend must have an [implementation in
+   OpenSAFELY Data Builder](data-builder-intro.md).
+
+## Implementing existing OpenSAFELY Contracts
+
+The existing [Contracts reference](contracts-reference.md) provides data
+specifications for both OpenSAFELY users and data providers.
+
+An OpenSAFELY backend implements one or more of these specifications. Each
+specification pertains to a specific health care data domain.
+
+For each data column in a Contract, the Contract details:
+
+* column name
+* data description
+* data type
+* any constraints
+
+Refer to those specifications when preparing data tables for integration
+with OpenSAFELY.
+
+### What if provided data does not exactly match the relevant Contract?
+
+Structuring data in line with OpenSAFELY Contracts makes it easier for
+researchers using OpenSAFELY to run studies across multiple data
+backends. However, there may be administrative or technical reasons why
+data providers cannot satisfy a Contract precisely.
+
+Some deviation from the Contract is permitted.
+
+TODO: What needs to be done if there is deviation? Is this following
+discussion and approval by OpenSAFELY?
+TODO: What counts as permitted deviation and what counts as a breach of
+the Contract?
```

Comment:
Alternate wording or clarification on this welcome. It's also OK to remove this section entirely for now, if we don't have this worked out yet.

Reply:
We don't know yet. For me, this is part of the point, rather than just some handwaving to be resolved in a month or two.

My hypothesis is that the process of agreeing on a contract implementation will often be a conversation, or even a negotiation, between the data provider, and the opensafely contract owner(s). This negotiation will over time cause us (the contract owners) to iterate our understanding of what constitutes a particular domain, and what concepts can reasonably be applied across different implementations of that domain.

So there will never really be a final version, although I'd expect core concepts / contracts to asymptotically approach a final version over time.

Does that make sense? I think it's probably worth attempting to capture some of the flavour of this in these docs.